superset/dataframe.py

```diff
@@ -41,26 +63,50 @@ class SupersetDataFrame(object):
         'V': None,   # raw data (void)
     }
 
-    def __init__(self, df):
-        self.__df = df.where((pd.notnull(df)), None)
+    def __init__(self, data, cursor_description, db_engine_spec):
+        column_names = []
+        if cursor_description:
+            column_names = [col[0] for col in cursor_description]
+        self.column_names = dedup(column_names)
+
+        # check whether the result set has any nested dict columns
+        if data:
+            has_dict_col = any([isinstance(c, dict) for c in data[0]])
+            df_data = list(data) if has_dict_col else np.array(data, dtype=object)
```

Comment:
Are you doing this because if a column has dicts all columns will have the `object` type in the dataframe when initialized from the numpy array? If so, why not use `infer_objects` here?

```python
df_data = np.array(data or [], dtype=object)
df = pd.DataFrame(df_data, columns=column_names).infer_objects()
```

Reply:
Just relocating the code here.

Reply:
I tested the behavior with `[(1, 2, {'a': 'b'})]` as data, trying to understand the code. If we build the dataframe using `np.array` on data **with** a dict we get the following types for the columns:

```python
>>> data = [(1, 2, {'a': 'b'})]
>>> df = pd.DataFrame(np.array(data, dtype=object))
>>> print(df.dtypes)
0    object
1    object
2    object
dtype: object
```

Here Pandas doesn't know that the first two columns are integers. OTOH, using `list(data)`:

```python
>>> data = [(1, 2, {'a': 'b'})]
>>> df = pd.DataFrame(list(data))
>>> print(df.dtypes)
0     int64
1     int64
2    object
dtype: object
```

Now we have correct type information for all columns. So it seems to me that the `list(data) if has_dict_col else np.array(data, dtype=object)` is trying to preserve the type information.

We can get the same result using `infer_objects()`:

```python
>>> data = [(1, 2, {'a': 'b'})]
>>> df = pd.DataFrame(np.array(data, dtype=object)).infer_objects()
>>> print(df.dtypes)
0     int64
1     int64
2    object
dtype: object
```

Reply:
Looks like the `dtype=object` was recently added here:
https://github.com/apache/incubator-superset/pull/4629
@michellethomas

Reply:
I added `dtype=object` just to solve an issue because np.array was erroring with data with different types in a tuple (`[('str', 1, [1], 1.0)]`). The `list(data) if...` logic was there already. `infer_objects` sounds like a good option though.

Reply:
Wow, looks like `infer_objects` does much of what Bogdan was trying to do later on in `dataframe.py`, maybe we can get ride of some of that code, though I hate to touch this as it's impossible to fully understand the implications... I'll read the source for it and decide if I can remove more code.